### DISCUSSION TITLE
Data leakage issue when using NeighborSampler?

### DISCUSSION THREAD

**Discussion by Dennis-Tsai** [pyg-team/pytorch_geometric#5059] [Q&A]
While using NeighborSampler, we set our selected `train_id` and `valid_id` (without overlapping) as the node_idx for creating the mini-batch training dataloaders as followed:

```
train_loader = NeighborSampler(
    edge_index,
    node_idx=train_idx,
    sizes=[-1, -1, -1],
    batch_size=64,
    shuffle=True,
    num_workers=12,
    drop_last=True,
)

val_loader = NeighborSampler(
    edge_index,
    node_idx=val_idx,
    sizes=[-1, -1, -1],
    batch_size=64,
    shuffle=True,
    num_workers=12,
    drop_last=True,
)
```

However, once we looped through the train_loader as shown below, the `n_id`, the nodes involved in the training computation, have some overlapping with our `val_idx`: 

```
for batch_size, n_id, adjs in train_loader:
    .....
```

The valid/test node data can be accessed in training process. I wonder if it is kinda data leakage problem? 
Please let me know if I misunderstood any part. Thanks!

**Comment by wsad1** [comment]
This is the expected behaviour.
The `n_id` is the id of all nodes involved in message passing this can overlap across train/val set since they can have  overlapping 3 hop neighbourhood. But only the `train_idx` nodes will be used for supervision.

**Reply by Dennis-Tsai** [reply]
Thanks for the prompt reply. I'm just thinking if there is any concern that the model could learned some knowledge from the valid dataset's node features in training process. In this case we could expect to gain a better performance when evaluating by this valid dataset.

Maybe adding a mask mechanism to prevent those overlapping would be more reasonable?

**Reply by rusty1s** [reply]
If you want to apply your model in an inductive rather than transductive learning scenario, you can simply filter it before passing it to the loader via `train_data = data.subgraph(data.train_mask)`.
